kolibri_zim_plugin/assets/src/views/ZimContentView.vue

```diff
@@ -25,66 +20,106 @@
+        this.loadArticle(path);
       },
-      onIframeLoad() {
-        try {
-          this.$refs.iframe.contentWindow.removeEventListener('unload', this.onIframeUnload);
-          this.$refs.iframe.contentWindow.addEventListener('unload', this.onIframeUnload);
-        } catch (DOMException) {
-          // pass
-        }
-        this.onIframeLocationChanged();
+      setContent(html, baseUrl) {
+        const parser = new DOMParser();
+        const document = parser.parseFromString(html, 'text/html');
+        // We can't set the base URI with shadow dom, so we will need to
+        // rewrite relative URLs...
+        document.getElementsByTagName('*').forEach(elem => {
+          if (elem.hasAttribute('href')) {
+            const href = elem.getAttribute('href');
+            const remapUrl = new URL(href, baseUrl);
+            elem.setAttribute('href', remapUrl);
+
+            const zimPath = this.zimPathFromUrl(remapUrl);
+
+            if (zimPath) {
+              elem.setAttribute('data-kolibri-zim-path', zimPath);
+              elem.setAttribute('data-kolibri-zim-hash', remapUrl.hash);
+            }
+
+            if (remapUrl.host !== baseUrl.host) {
+              elem.setAttribute('target', '_blank');
+            }
+          } else if (elem.hasAttribute('src')) {
+            const src = elem.getAttribute('src');
+            const remapUrl = new URL(src, baseUrl);
+            elem.setAttribute('src', remapUrl);
+          }
+        });
+        this.shadow.appendChild(document.documentElement);
+        return document;
       },
-      onIframeUnload() {
-        setTimeout(this.onIframeLocationChanged, 0);
-      },
-      onIframeLocationChanged() {
-        // FIXME: Building the breadcrumb trail here involves reading from
-        //        iframe.contentWindow.location, which is not possible with
-        //        cross-origin iframes. This will need to be rewritten when
-        //        the zim backend is moved to a different server.
-
-        let href, title, isLoading, isExternal;
-        const contentDocument = this.$refs.iframe.contentDocument;
-        const contentWindow = this.$refs.iframe.contentWindow;
-
-        try {
-          href = contentWindow.location.href;
-        } catch (DOMException) {
-          href = undefined;
-        }
+      zimPathFromUrl(url) {
+        const fullPath = url.pathname;
+        const basePath = urls.zim_article(this.zimFilename, '');
 
-        if (contentDocument && contentDocument.readyState == 'loading') {
-          title = '…';
-          isLoading = true;
-        } else if (contentDocument) {
-          title = contentDocument.title;
+        if (fullPath.startsWith(basePath)) {
+          return fullPath.substr(basePath.length);
         } else {
-          title = '…';
-          isExternal = true;
+          return undefined;
+        }
+      },
+      onShadowClick(event) {
+        const linkElem = event.originalTarget.closest('a');
```

Comment:
It looks like this is only working on Firefox: https://developer.mozilla.org/en-US/docs/Web/API/Event/originalTarget

Reply:
Oh, good spotting! I replaced this with the following, which appears to be the most widely compatible way of doing what we need here:

```
const clickElem = this.shadow.elementFromPoint(event.clientX, event.clientY);
const linkElem = clickElem ? clickElem.closest('a') : undefined;
```